File: tests/test_core_logic.py
"""Regression tests for the P0/P1 bugs fixed in the audit.

These tests do not touch the network. _download_file and the graphics-tools
helpers are monkeypatched to return controlled fixtures.
"""
import json
import pathlib
import sys
import tarfile

import pytest

# Make the repo root importable when pytest is invoked from anywhere.
sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent.parent))

import core_logic as core


# --- parse_package_name_from_tgz -------------------------------------------------

def test_parse_handles_graphics_tools_prerelease():
    name, ver = core.parse_package_name_from_tgz(
        "com.microsoft.mrtk.graphicstools.unity-1.0.0-pre.1.tgz"
    )
    assert name == "com.microsoft.mrtk.graphicstools.unity"
    assert ver == "1.0.0-pre.1"


def test_parse_handles_plain_mrtk_release():
    name, ver = core.parse_package_name_from_tgz("org.mixedrealitytoolkit.core-3.0.0.tgz")
    assert name == "org.mixedrealitytoolkit.core"
    assert ver == "3.0.0"


def test_parse_rejects_unrelated_filename():
    assert core.parse_package_name_from_tgz("README.md") == (None, None)
    assert core.parse_package_name_from_tgz("foo.tgz") == (None, None)


# --- UnityVersion ----------------------------------------------------------------

def test_unity_version_parser_handles_malformed_strings():
    # Should not raise on any of these inputs.
    for bad in ["", "garbage", "1.x.0f0", "1.0", None]:
        try:
            core.UnityVersion(bad)
        except Exception as e:
            pytest.fail(f"UnityVersion({bad!r}) raised {e!r}")


def test_unity_version_ordering():
    assert core.UnityVersion("6000.0.0f0") > core.UnityVersion("2022.3.0f1")
    assert core.UnityVersion("2022.3.0f1") > core.UnityVersion("2022.2.10f1")
    assert core.UnityVersion("2022.3.0f1") == core.UnityVersion("2022.3.0f1")
    assert not (core.UnityVersion("2022.3.0f1") > core.UnityVersion("2022.3.0f1"))


# --- Dependency resolution -------------------------------------------------------

def _make_fake_release(filename, url="https://example.invalid/asset"):
    return {"assets": [{"name": filename, "browser_download_url": url}]}


def _write_fake_tgz(path: pathlib.Path, package_json: dict):
    """Write a minimal .tgz containing only `package/package.json`."""
    import io
    payload = json.dumps(package_json).encode()
    with tarfile.open(path, "w:gz") as tar:
        info = tarfile.TarInfo(name="package/package.json")
        info.size = len(payload)
        tar.addfile(info, io.BytesIO(payload))


def test_dependency_resolution_terminates_on_cycle(monkeypatch, tmp_path):
    """A → B → A loop must not infinite-loop."""
    releases = [
        _make_fake_release("org.mixedrealitytoolkit.a-1.0.0.tgz", "https://example.invalid/a"),
        _make_fake_release("org.mixedrealitytoolkit.b-1.0.0.tgz", "https://example.invalid/b"),
    ]

    def fake_download(url, dest_folder, progress_callback=None, dest_name=None):
        fname = dest_name or url.split("/")[-1] + ".tgz"
        dest = pathlib.Path(dest_folder) / fname
        if url.endswith("/a"):
            _write_fake_tgz(dest, {"dependencies": {"org.mixedrealitytoolkit.b": "1.0.0"}})
        else:
            _write_fake_tgz(dest, {"dependencies": {"org.mixedrealitytoolkit.a": "1.0.0"}})
        return dest

    monkeypatch.setattr(core, "_download_file", fake_download)

    resolved = core.resolve_dependencies({"a": "1.0.0"}, releases, work_dir=tmp_path)
    assert "a" in resolved
    assert "b" in resolved


def test_graphics_tools_is_recorded_when_pulled_in_as_dep(monkeypatch, tmp_path):
    """Graphics-tools is fetched from its own repo at download time and has no
    release asset in the MRTK repo. When an MRTK package declares it as a dep,
    it must still be recorded in resolved_packages so download_and_apply_packages
    knows to repackage it. (Parity with C++ behaviour: record, don't recurse.)"""
    releases = [
        _make_fake_release("org.mixedrealitytoolkit.foo-1.0.0.tgz", "https://example.invalid/foo"),
    ]

    def fake_download(url, dest_folder, progress_callback=None, dest_name=None):
        fname = dest_name or url.split("/")[-1]
        dest = pathlib.Path(dest_folder) / fname
        # `foo` declares graphics-tools as a dependency.
        _write_fake_tgz(dest, {"dependencies": {core.GRAPHICS_TOOLS_NAME: "1.5.0"}})
        return dest

    monkeypatch.setattr(core, "_download_file", fake_download)
    resolved = core.resolve_dependencies({"foo": "1.0.0"}, releases, work_dir=tmp_path)
    assert "foo" in resolved
    assert core.GRAPHICS_TOOLS_NAME in resolved
    assert resolved[core.GRAPHICS_TOOLS_NAME] == "1.5.0"


# --- Install side effects --------------------------------------------------------

def _make_fake_unity_project(root: pathlib.Path):
    root.mkdir(parents=True, exist_ok=True)
    (root / "Assets").mkdir()
    (root / "Packages").mkdir()
    (root / "ProjectSettings").mkdir()
    (root / "Packages" / "manifest.json").write_text(json.dumps({"dependencies": {}}))
    return root


def test_install_preserves_existing_mixed_reality_folder(monkeypatch, tmp_path):
    """Issue 3: do NOT rmtree <project>/Packages/MixedReality before writing."""
    project = _make_fake_unity_project(tmp_path / "proj")
    mr = project / "Packages" / "MixedReality"
    mr.mkdir()
    preexisting = mr / "preexisting.tgz"
    preexisting.write_text("untouched")

    work_dir = tmp_path / "work"
    work_dir.mkdir()

    new_pkg = work_dir / "org.mixedrealitytoolkit.core-3.0.0.tgz"
    new_pkg.write_text("new")

    def fake_download(url, dest_folder, progress_callback=None, dest_name=None):
        return new_pkg

    monkeypatch.setattr(core, "_download_file", fake_download)

    releases = [_make_fake_release("org.mixedrealitytoolkit.core-3.0.0.tgz",
                                    "https://example.invalid/core")]
    ok = core.download_and_apply_packages(
        project, {"core": "3.0.0"}, releases, set(), work_dir=work_dir
    )
    assert ok
    assert preexisting.exists(), "Pre-existing MixedReality file was destroyed"
    assert preexisting.read_text() == "untouched"
    assert (mr / "org.mixedrealitytoolkit.core-3.0.0.tgz").exists()


def test_install_ignores_unrelated_tgz_in_cwd(monkeypatch, tmp_path):
    """Issue 2: install must NOT glob the CWD/work_dir for tgz files. Only files
    actually produced by _download_file should end up in the project."""
    project = _make_fake_unity_project(tmp_path / "proj")
    work_dir = tmp_path / "work"
    work_dir.mkdir()

    # Plant an unrelated tgz in work_dir — install must not pick it up.
    unrelated = work_dir / "unrelated-other-1.0.0.tgz"
    unrelated.write_text("not mine")

    new_pkg = work_dir / "org.mixedrealitytoolkit.core-3.0.0.tgz"
    new_pkg.write_text("new")

    def fake_download(url, dest_folder, progress_callback=None, dest_name=None):
        return new_pkg

    monkeypatch.setattr(core, "_download_file", fake_download)

    releases = [_make_fake_release("org.mixedrealitytoolkit.core-3.0.0.tgz",
                                    "https://example.invalid/core")]
    core.download_and_apply_packages(project, {"core": "3.0.0"}, releases, set(), work_dir=work_dir)

    installed = list((project / "Packages" / "MixedReality").iterdir())
    installed_names = {p.name for p in installed}
    assert "org.mixedrealitytoolkit.core-3.0.0.tgz" in installed_names
    assert "unrelated-other-1.0.0.tgz" not in installed_names

    manifest = json.loads((project / "Packages" / "manifest.json").read_text())
    assert "org.mixedrealitytoolkit.core" in manifest["dependencies"]
    # The unrelated file's name must not have leaked into the manifest either.
    assert "unrelated-other" not in manifest["dependencies"]

File: tests/test_core_logic.cpp
// Lightweight regression harness for P0/P1 bugs fixed in the audit.
// No external test framework — failures produce non-zero exit code.

#include "core_logic.h"

#include <cassert>
#include <cstdio>
#include <fstream>
#include <iostream>
#include <string>

namespace {
int g_failures = 0;

#define EXPECT(cond, msg)                                              \
    do {                                                               \
        if (!(cond)) {                                                 \
            std::cerr << "FAIL: " << __func__ << ":" << __LINE__       \
                      << " " << (msg) << " — `" #cond "` was false\n"; \
            ++g_failures;                                              \
        }                                                              \
    } while (0)

void test_parse_handles_graphics_tools_prerelease() {
    auto [name, ver] = parsePackageNameFromTgz("com.microsoft.mrtk.graphicstools.unity-1.0.0-pre.1.tgz");
    EXPECT(name == "com.microsoft.mrtk.graphicstools.unity", "graphics-tools name");
    EXPECT(ver  == "1.0.0-pre.1", "graphics-tools version");
}

void test_parse_handles_plain_mrtk_release() {
    auto [name, ver] = parsePackageNameFromTgz("org.mixedrealitytoolkit.core-3.0.0.tgz");
    EXPECT(name == "org.mixedrealitytoolkit.core", "mrtk name");
    EXPECT(ver  == "3.0.0", "mrtk version");
}

void test_parse_rejects_garbage() {
    auto [name, ver] = parsePackageNameFromTgz("README.md");
    EXPECT(name.empty(), "garbage filename → empty name");
    EXPECT(ver.empty(),  "garbage filename → empty version");
    auto [n2, v2] = parsePackageNameFromTgz("foo.tgz");
    EXPECT(n2.empty(), "no-version tgz → empty name");
}

void test_unity_version_parser_handles_malformed() {
    // None of these should throw.
    try {
        UnityVersion v1("");
        UnityVersion v2("garbage");
        UnityVersion v3("1.x.0f0");
        UnityVersion v4("1.0");
        (void)v1; (void)v2; (void)v3; (void)v4;
    } catch (const std::exception& e) {
        EXPECT(false, std::string("UnityVersion ctor threw: ") + e.what());
    }
}

void test_unity_version_ordering() {
    UnityVersion u6{"6000.0.0f0"};
    UnityVersion u22{"2022.3.0f1"};
    UnityVersion u22b{"2022.3.0f1"};
    UnityVersion u_older{"2022.2.10f1"};
    EXPECT(u6 > u22, "6000 > 2022.3");
    EXPECT(u22 > u_older, "2022.3 > 2022.2.10");
    EXPECT(!(u22 > u22b), "equal versions not >");
}

void test_get_unity_version_handles_malformed_line() {
    // Smoke-check using a temp ProjectVersion.txt to confirm regex parser
    // returns the expected value and does not regress to returning the whole line.
    namespace fs = std::filesystem;
    fs::path tmp = fs::temp_directory_path() / "mrtk_test_unity_proj";
    fs::create_directories(tmp / "ProjectSettings");
    {
        std::ofstream f(tmp / "ProjectSettings" / "ProjectVersion.txt");
        f << "m_EditorVersion: 2022.3.10f1\n";
        f << "m_EditorVersionWithRevision: 2022.3.10f1 (abc123)\n";
    }
    std::string v = MRTKToolCore::getUnityVersion(tmp);
    EXPECT(v == "2022.3.10f1", std::string("expected 2022.3.10f1, got '") + v + "'");
    fs::remove_all(tmp);
}
} // namespace

int main() {
    test_parse_handles_graphics_tools_prerelease();
    test_parse_handles_plain_mrtk_release();
    test_parse_rejects_garbage();
    test_unity_version_parser_handles_malformed();
    test_unity_version_ordering();
    test_get_unity_version_handles_malformed_line();

    if (g_failures == 0) {
        std::cout << "All tests passed.\n";
        return 0;
    }
    std::cerr << g_failures << " test failure(s).\n";
    return 1;
}
